src/System.Private.CoreLib/shared/System/Number.DiyFp.cs

```diff
@@ -8,121 +8,141 @@ namespace System
-        internal struct DiyFp
+        // This is a port of the `DiyFp` implementation here: https://github.com/google/double-conversion/blob/a711666ddd063eb1e4b181a6cb981d39a1fc8bac/double-conversion/diy-fp.h
+        // The backing structure and how it is used is described in more detail here: http://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf
+
+        // This "Do It Yourself Floating Point" class implements a floating-point number with a ulong significand and an int exponent.
+        // Normalized DiyFp numbers will have the most significant bit of the significand set.
+        // Multiplication and Subtraction do not normalize their results.
+        // DiyFp are not designed to contain special doubles (NaN and Infinity).
+        internal readonly ref struct DiyFp
         {
-            public const int SignificandLength = 64;
+            public const int SignificandSize = 64;
 
-            // Extended significand.
-            // IEEE 754 double-precision numbers only require 53 bits significand.
-            // However, in Grisu3 we need additional 11 bits so we declare _f as ulong.
-            // Please note _f does not include sign bit.
-            private ulong _f;
+            public readonly ulong f;
+            public readonly int e;
 
-            // Extended exponent.
-            // IEEE 754 double-precision numbers only require 11 bits exponent.
-            // However, in Grisu3 we need extra space to avoid overflow so we declare _e as int.
-            // Please note _e is a biased exponent if the DiyFp instance is generated by GenerateNormalizedDiyFp().
-            private int _e;
-
-            public DiyFp(ulong f, int e)
+            public DiyFp(double value)
             {
-                _f = f;
-                _e = e;
+                Debug.Assert(double.IsFinite(value));
+                Debug.Assert(value > 0.0);
+                f = ExtractFractionAndBiasedExponent(value, out e);
             }
 
-            public ulong f
+            public DiyFp(float value)
             {
-                get
-                {
-                    return _f;
-                }
+                Debug.Assert(float.IsFinite(value));
+                Debug.Assert(value > 0.0f);
+                f = ExtractFractionAndBiasedExponent(value, out e);
+            }
 
-                set
-                {
-                    _f = value;
-                }
+            // Computes the two boundaries of value.
+            //
+            // The bigger boundary (mPlus) is normalized.
+            // The lower boundary has the same exponent as mPlus.
+            //
+            // Precondition:
+            //  The value encoded by value must be greater than 0.
+            public DiyFp(double value, out DiyFp mMinus, out DiyFp mPlus) : this(value)
```

Comment:
`out` parameters in a constructor feels odd to me. Would it make more sense to make this a static method? Or even expose `GetBoundaries` on the type?

Reply:
> Would it make more sense to make this a static method? Or even expose GetBoundaries on the type?

We might be able to make it a static method (`CreateValueAndBoundaries` or something), but it would involve duplicating the contents of the existing `DiyFp` constructor

We don't want to expose `GetBoundaries` publicly because it is only valid to get the boundaries (using the given algorithm) on a non-normalized `DiyFp`

Reply:
> but it would involve duplicating the contents of the existing DiyFp constructor

How? Why couldn't the method by implemented as:

```C#
static DiyFp CreateValueAndBoundaries(double value, out DiyFp mMinus, out DiyFp mPlus)
{
DiyFp result = new DiyFp(value);
result.GetBoundaries(52, out mMinus, out mPlus);
return result;
}
```

---
In reply to: [252368004](https://github.com/dotnet/coreclr/pull/22040#discussion_r252368004) [](ancestors = 252368004)

Reply:
You're right, will update.

Reply:
Fixed. Method was named `CreateAndGetBoundaries`